javacutil/src/main/java/org/checkerframework/javacutil/TreeUtils.java

```diff
@@ -355,45 +354,53 @@ public static ExecutableElement elementFromUse(NewClassTree node) {
   }
 
   /**
-   * Determines the symbol for a constructor given an invocation via {@code new}.
+   * Returns the super constructor invoked by {@code newClassTree}.
+   *
+   * <p>If {@code newClassTree} is an anonymous class, then this method returns the super
+   * constructor invoked in the body of the anonymous constructor.
    *
-   * <p>If the tree is a declaration of an anonymous class, then method returns constructor that
-   * gets invoked in the extended class, rather than the anonymous constructor implicitly added by
-   * the constructor (JLS 15.9.5.1)
+   * <p>If {@code newClassTree} is not an anonymous class, then this method returns {@link
+   * #constructor(NewClassTree)}.
+   *
+   * @param newClassTree the constructor invocation
+   * @return the super constructor invoked in the body of the anonymous constructor; or {@link
+   *     #constructor(NewClassTree)} if {@code newClassTree} is not an anonymous class
+   */
+  public static ExecutableElement getSuperConstructor(NewClassTree newClassTree) {
+    if (newClassTree.getClassBody() == null) {
+      return constructor(newClassTree);
+    }
+    JCNewClass jcNewClass = (JCNewClass) newClassTree;
+    // anonymous constructor bodies should contain exactly one statement
+    // in the form:
+    //    super(arg1, ...)
+    // or
+    //    o.super(arg1, ...)
+    //
+    // which is a method invocation of the super constructor
+
+    // the method call is guaranteed to return nonnull
+    JCMethodDecl anonConstructor =
+        (JCMethodDecl) TreeInfo.declarationFor(jcNewClass.constructor, jcNewClass);
+    assert anonConstructor != null;
+    assert anonConstructor.body.stats.size() == 1;
+    JCExpressionStatement stmt = (JCExpressionStatement) anonConstructor.body.stats.head;
+    JCMethodInvocation superInvok = (JCMethodInvocation) stmt.expr;
+    return (ExecutableElement) TreeInfo.symbol(superInvok.meth);
+  }
+
+  /**
+   * Determines the symbol for a constructor given an invocation via {@code new}.
    *
    * @see #elementFromUse(NewClassTree)
    * @param tree the constructor invocation
    * @return the {@link ExecutableElement} corresponding to the constructor call in {@code tree}
    */
   public static ExecutableElement constructor(NewClassTree tree) {
-
-    if (!(tree instanceof JCTree.JCNewClass)) {
+    if (!(tree instanceof JCNewClass)) {
       throw new BugInCF("TreeUtils.constructor: not a javac internal tree");
```

Comment:
Is this necessary?  The next line does a cast, so the error message will be essentially the same with or without this `if` and `throw`.